src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -984,6 +984,253 @@ private Task WriteSuffix()
+            {
+                unsafe
+                {
+                    // Multiple buffers and < stackAllocLimit, copy into a stack buffer
+                    byte* stackBuffer = stackalloc byte[stackAllocLimit];
+                    span = new Span<byte>(stackBuffer, startLineBuffer.Length);
+                    startLineBuffer.CopyTo(span);
+                }
+            }
+            else
+            {
+                // We're not a single span here but we can use pooled arrays to avoid allocations in the rare case
+                span = new Span<byte>(new byte[startLineBuffer.Length]);
+                startLineBuffer.CopyTo(span);
+            }
+
+            var methodEnd = 0;
+            if (!span.GetKnownMethod(out string method))
+            {
+                methodEnd = span.IndexOf(ByteSpace);
+                if (methodEnd == -1)
+                {
+                    RejectRequestLine(start, end);
+                }
+
+                method = span.Slice(0, methodEnd).GetAsciiString();
+
+                if (method == null)
+                {
+                    RejectRequestLine(start, end);
+                }
+
+                // Note: We're not in the fast path any more (GetKnownMethod should have handled any HTTP Method we're aware of)
+                // So we can be a tiny bit slower and more careful here.
+                for (int i = 0; i < method.Length; i++)
+                {
+                    if (!IsValidTokenChar(method[i]))
+                    {
+                        RejectRequestLine(start, end);
+                    }
+                }
+            }
+            else
+            {
+                methodEnd += method.Length;
+            }
+
+            var needDecode = false;
+            var pathBegin = methodEnd + 1;
+            var pathToEndSpan = span.Slice(pathBegin, span.Length - pathBegin);
+            // var pbs = Encoding.UTF8.GetString(pathToEndSpan.ToArray());
+            pathBegin = 0;
+
+            // TODO: IndexOfAny
+            var spaceIndex = pathToEndSpan.IndexOf(ByteSpace);
+            var questionMarkIndex = pathToEndSpan.IndexOf(ByteQuestionMark);
+            var percentageIndex = pathToEndSpan.IndexOf(BytePercentage);
+
+            var pathEnd = MinNonZero(spaceIndex, questionMarkIndex, percentageIndex);
+
+            if (spaceIndex == -1 && questionMarkIndex == -1 && percentageIndex == -1)
+            {
+                RejectRequestLine(start, end);
+            }
+            else if (percentageIndex != -1)
+            {
+                needDecode = true;
+
+                pathEnd = MinNonZero(spaceIndex, questionMarkIndex);
+                if (questionMarkIndex == -1 && spaceIndex == -1)
+                {
+                    RejectRequestLine(start, end);
+                }
+            }
+
+            var queryString = "";
+            var queryEnd = pathEnd;
+            if (questionMarkIndex != -1)
```

Comment:
I need to run the tests to see if this even works. I'm pretty sure I botched the logic here 😄